### PR TITLE
[STOR-513]: Support for restore using local snapshots in generic backup.

### PR DESCRIPTION
**Pull request by diptiranjanpx**
**What type of PR is this?**
> Uncomment only one and also add the corresponding label in the PR:
>bug
feature
>cleanup
>api-change
>design
>documentation
>failing-test
>unit-test
>integration-test

**What this PR does / why we need it**:
Support for restore using local snapshots in generic backup if available.

**Does this PR change a user-facing CRD or CLI?**:

**Is a release note needed?**:

```release-note

```

**Does this change need to be cherry-picked to a release branch?**:

2 restores , one was latest which was restored using local snapshot and the other prev one was restored as a generic restore.

Screenshot 2021-11-10 at 6 13 56 PM

### CODE REVIEW

**Review by siva-portworx — APPROVED** [review]

**Review by prashanthpx — APPROVED** [review]
lgtm

File: drivers/volume/kdmp/kdmp.go [hunk 1]
```diff
@@ -718,9 +744,6 @@ func (k *kdmp) GetRestoreStatus(restore *storkapi.ApplicationRestore) ([]*storka
 				vInfo.Reason = "Volume restore is in progress. BytesDone"
 				// TODO: KDMP does not return size right now, we will need to fill up
 				// size for volume once support is available
-			} else if dataExport.Status.Status == kdmpapi.DataExportStatusFailed {
-				vInfo.Status = storkapi.ApplicationRestoreStatusFailed
-				vInfo.Reason = fmt.Sprintf("restore failed for volume:%s, %s", vInfo.SourceVolume, dataExport.Status.Reason)
 			} else if isDataExportCompleted(dataExport.Status) {
 				restoredPVC, err := core.Instance().GetPersistentVolumeClaim(dataExport.Status.RestorePVC.Name, dataExport.Status.RestorePVC.Namespace)
 				if err != nil {
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply drivers/volume/kdmp/kdmp.go:749]: This is in earlier part of the code where we had moved the Failure check with more condition like stage should have been final. There the reason of failure is updated as part of each volume info.

File: drivers/volume/kdmp/kdmp.go [hunk 2]
```diff
@@ -718,9 +744,6 @@ func (k *kdmp) GetRestoreStatus(restore *storkapi.ApplicationRestore) ([]*storka
 				vInfo.Reason = "Volume restore is in progress. BytesDone"
 				// TODO: KDMP does not return size right now, we will need to fill up
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply drivers/volume/kdmp/kdmp.go:745]: done

File: pkg/snapshotter/snapshotter_csi.go [hunk 3]
```diff
@@ -852,3 +877,112 @@ func (c *csiDriver) RetainLocalSnapshots(
...
+	var vsCRPath string
+
+	vsCRList, err := c.getCSISnapshotsCRList(backupLocation, pvcUID, objectPath)
+	if err != nil {
+		return snapshotInfo, fmt.Errorf("failed in getting list of older volumesnapshot CRs from objectstore : %v", err)
+	}
+	for _, volumeSnapshotCR := range vsCRList {
+		tempBackupUID, _ := getBackupInfoFromObjectKey(volumeSnapshotCR)
+		if tempBackupUID == backupUID {
+			logrus.Debugf("resources of local snapshot for backup %s are prsent in the objectpath %s", backupUID, objectPath)
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/snapshotter/snapshotter_csi.go:893]: done

File: drivers/volume/kdmp/kdmp.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (siva-portworx) [reply drivers/volume/kdmp/kdmp.go:757]: Actually Prashanth comment is valid, currently we will be missing the failure reason for the volumes, if any. We can fix it in next PR. I am merge this PR for now.

**Pull request merged**